### PR TITLE
feat(1967): add ability to do raw query

### PR DESCRIPTION
**Pull request by InderH**
## Context

Need to add the ability to do raw queries to baseFactory so queries that are otherwise impossible to do through sequelize can be done. Need to add ability to get build status history for jobs in bulk so that they can be displayed in pipeline list view.

## Objective

Add query method to baseFactory. Add getBuildStatuses to buildFactory so build status history for jobs can be queried in bulk.

## References

https://github.com/screwdriver-cd/screwdriver/issues/1967

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: lib/baseFactory.js [hunk 1]
```diff
@@ -220,6 +220,51 @@ class BaseFactory {
...
+            datastoreForLookup = this.datastoreRO;
+        }
+
+        return datastoreForLookup.query(queryConfig)
+            .then((data) => {
+                if (queryConfig.rawResponse) {
+                    return data;
+                }
+
+                const result = [];
```
Comment (pritamstyz4ever) [inline lib/baseFactory.js]: Can we use .map() instead of doing the for each?

File: lib/buildFactory.js [hunk 2]
```diff
@@ -295,6 +295,72 @@ class BuildFactory extends BaseFactory {
...
+     * @param  {Number}   config.offset           Number of build statuses to skip
+     * @param  {Number}   config.numBuilds        Number of build statuses to return per job
+     * @return {Promise}
+     */
+    getBuildStatuses(config) {
+        config.offset = config.offset || 0;
+        config.numBuilds = config.numBuilds || 1;
+        config.jobIds = config.jobIds || [];
+
+        const query =
```
Comment (pritamstyz4ever) [inline lib/buildFactory.js]: Use string literals instead of concatenating the query

Comment (pritamstyz4ever) [reply lib/buildFactory.js]: Also, thinking would it make sense to have a separate file for raw queries which can be required here. Would be easy to compare when we do several changes over the time and also more maintainable.

Comment (InderH) [reply lib/buildFactory.js]: If I do a string literal the string exceeds the maximum lint line length, when I tried to do a multi line string I got another lint issue about multi line strings only being supported in es6+.

Is there another way to do this or should I disable the linter for the string?

**Review by parthasl — DISMISSED** [review]
👍

File: lib/rawQueries.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+        WHERE rank > :offset AND rank <= :maxRank
+        ORDER BY r."jobId", r."id" DESC`,
+
+    statusesQueryMySql:
+        `SELECT r.id, r.jobId, j.name as \`jobName\`, r.status, r.startTime, r.endTime
+        FROM (SELECT id, jobId, status, startTime, endTime,
+        RANK() OVER ( PARTITION BY jobId ORDER BY id DESC ) AS \`rank\`
+        FROM builds WHERE jobId in (:jobIds)) r
+        INNER JOIN jobs as j ON j.id = r.jobId
+        WHERE \`rank\` > :offset AND \`rank\` <= :maxRank
```
Comment (parthasl) [inline lib/rawQueries.js:15]: @InderH for mysql we need ` prefix ?

Comment (parthasl) [reply lib/rawQueries.js:15]: ignore, got it

**Comment by parthasl** [comment]
@InderH - may I know the frequency of this query? this read-only query will be routed to RO instance?

**Comment by InderH** [comment]
> @InderH - may I know the frequency of this query? this read-only query will be routed to RO instance?

@parthasl - This query will run when:
1. the list view is loaded
2. when the list view is scrolled down and there is more to load
3. when the list view's refresh button is pressed

Yes, the query is done with readOnly set to true, so it uses the RO db
https://github.com/screwdriver-cd/models/pull/441/files#diff-43e1dd855c9ebdae88cbf0102c6e05b2R330
https://github.com/screwdriver-cd/models/pull/441/files#diff-c6dfbb25b49dd3e823a2aa3a6f797f96R249

File: lib/baseFactory.js [hunk 4]
```diff
@@ -223,6 +223,48 @@ class BaseFactory {
...
+        let datastoreForLookup = this.datastore;
+        // use read only datastore if it is configured
+        const readOnly = hoek.reach(config, 'readOnly', { default: false });
+
+        if (readOnly) {
+            datastoreForLookup = this.datastoreRO;
+        }
+
+        return datastoreForLookup.query(queryConfig).then(data => {
+            let toReturn = data;
```
Comment (tkyi) [inline lib/baseFactory.js]: maybe a better var name than `toReturn`?

**Review by tkyi — CHANGES_REQUESTED** [review]
can you update https://github.com/screwdriver-cd/models/blob/master/README.md#build-factory

File: lib/baseFactory.js (hunk 4 above)
Comment (InderH) [reply lib/baseFactory.js]: What do you think of something like 'formattedData'?

Comment (adong) [reply lib/baseFactory.js]: Good call @tkyi 
Yeah, @InderH that `formattedData` would make more sense.

**Review by tkyi — APPROVED** [review]

File: lib/buildFactory.js [hunk 5]
```diff
@@ -306,6 +307,51 @@ class BuildFactory extends BaseFactory {
...
+     * Get build statuses for jobs
+     * @method getBuildStatuses
+     * @param  {Object}   config                  Config object
+     * @param  {Array}    config.jobIds           Jobs to get build statuses for
+     * @param  {Number}   config.offset           Number of build statuses to skip
+     * @param  {Number}   config.numBuilds        Number of build statuses to return per job
+     * @return {Promise}
+     */
+    getBuildStatuses(config) {
+        config.offset = config.offset || 0;
```
Comment (tkyi) [inline lib/buildFactory.js]: probably shouldn't override function params like this
might be better to create a new var like this:
```suggestion
        const offset = config.offset || 0;
```

Comment (InderH) [reply lib/buildFactory.js]: Done, thanks!

**Review by tkyi — DISMISSED** [review]
👍

File: lib/baseFactory.js [hunk 6]
```diff
@@ -223,6 +223,48 @@ class BaseFactory {
...
+            rawResponse: hoek.reach(config, 'rawResponse', { default: false })
+        };
+
+        let datastoreForLookup = this.datastore;
+        // use read only datastore if it is configured
+        const readOnly = hoek.reach(config, 'readOnly', { default: false });
+
+        if (readOnly) {
+            datastoreForLookup = this.datastoreRO;
+        }
```
Comment (pritamstyz4ever) [inline lib/baseFactory.js]:
```suggestion
        // use read only datastore if it is configured
        const readOnly = hoek.reach(config, 'readOnly', { default: false });
        let datastoreForLookup = readOnly ? this.datastoreRO : this.datastore;
```

**Review by pritamstyz4ever — DISMISSED** [review]

File: lib/baseFactory.js (hunk 6 above)
Comment (InderH) [reply lib/baseFactory.js]: Done, thanks!

File: test/lib/buildFactory.test.js [hunk 7]
```diff
@@ -1005,4 +1007,175 @@ describe('Build Factory', () => {
...
+                offset: 1,
+                numBuilds: 5
+            };
+
+            returnValue = [
+                [
+                    {
+                        jobId: 1,
+                        jobName: 'name',
+                        Status: 'SUCCESS',
```
Comment (tkyi) [inline test/lib/buildFactory.test.js]: should be lowercase
```suggestion
                        status: 'SUCCESS',
```

File: test/lib/buildFactory.test.js [hunk 8]
```diff
@@ -1005,4 +1007,175 @@ describe('Build Factory', () => {
...
+            sinon.stub(BuildFactory.prototype, 'query').returns();
+
+            config = {
+                jobIds: [1, 2, 3, 4],
+                offset: 1,
+                numBuilds: 5
+            };
+
+            returnValue = [
+                [
```
Comment (tkyi) [inline test/lib/buildFactory.test.js:1027]: also is this the correct return data type? an array of arrays?

File: test/lib/buildFactory.test.js (hunk 7 above)
Comment (adong) [reply test/lib/buildFactory.test.js]: I wonder this too.

File: test/lib/buildFactory.test.js (hunk 8 above)
Comment (adong) [reply test/lib/buildFactory.test.js:1027]: @InderH

File: test/lib/buildFactory.test.js (hunk 7 above)
Comment (InderH) [reply test/lib/buildFactory.test.js]: updated, thanks!

File: test/lib/buildFactory.test.js (hunk 8 above)
Comment (InderH) [reply test/lib/buildFactory.test.js:1027]: yup, it should be an array of arrays - when running a raw query on sequelize without having it map the returning data to a model (which in this case we don't), it returns an array with 2 elements, the first being an array of the data/rows, and the second being meta data.

Comment (tkyi) [reply test/lib/buildFactory.test.js:1027]: hm but the expected query config has `table: 'builds'`, right?

Comment (InderH) [reply test/lib/buildFactory.test.js:1027]: Yeah it has `table: 'builds'` but also has `rawResponse: true` - we are only getting a small subset of the data a build has and are using the data in a way which casting it to a build model wouldn't help, so having the query response map back to builds model would just be extra unnecessary computation

Comment (tkyi) [reply test/lib/buildFactory.test.js:1027]: I see
sounds reasonable

**Review by tkyi — APPROVED** [review]

File: lib/rawQueries.js (hunk 3 above)
Comment (adong) [reply lib/rawQueries.js:15]: I just used a formatter to read, its essentially 
```sql
SELECT "id", 
       "jobid", 
       "status", 
       "starttime", 
       "endtime" 
FROM   (SELECT "id", 
               "jobid", 
               "status", 
               "starttime", 
               "endtime", 
               Rank() 
                 OVER ( 
                   partition BY "jobid" 
                   ORDER BY "id" DESC ) AS rank 
        FROM   builds 
        WHERE  "jobid" IN ( :jobIds )) AS R 
WHERE  rank > :offset 
       AND rank <= :maxRank 
ORDER  BY "jobid", 
          "id" DESC 
```

**Pull request merged**